### PR TITLE
[WIP] Explode on info.xml problems

### PR DESCRIPTION
**Pull request by butonic**
fixing https://github.com/owncloud/core/issues/30190

based on https://github.com/owncloud/core/pull/30188

cc @Tom @VicDeo @PVince81 lets see what breaks

### CODE REVIEW

File: lib/private/App/InfoParser.php [hunk 1]
```diff
@@ -30,10 +30,12 @@ class InfoParser {
 	/**
 	 * @param string $file the xml file to be loaded
 	 * @return null|array where null is an indicator for an error
+	 * @throws \NotFoundException oif file does not exist
```
Comment (patrickjahns) [inline lib/private/App/InfoParser.php]: typo

**Comment by butonic** [comment]
@VicDeo @PVince81 @tomneedham I had a look at this ... a lot of places need touching if we want to make oc explode if an app contains invalid XML. Also apps that no longer exist are a problem because the version reported for them is `0` which AFAIU is the problem for migrations not happening ...

Also, should we introduce a new axcaptions not existing on disk or should we reuse `OCP\App\AppNotFoundException`? The phpdoc indicates it is for apps not found in the market

feel free to highjack this PR. I doubt I will get to it during the week. I had hoped it would be a smaller task and I wanted to build it upon https://github.com/owncloud/core/pull/30188 befor @VicDeo might run off in a direction that would conflict with the changes in that PR ...

**Comment by PVince81** [comment]
is all the etag stuff really necessary just for throwing exceptions on error ?

File: lib/private/legacy/app.php [hunk 2]
```diff
@@ -616,44 +626,122 @@ public static function getAppVersionByPath($path) {
...
+			return "${stat['mtime']}|${stat['ino']}|${stat['dev']}|${stat['size']}";
+		}
+		return false;
+	}
+
+	/**
+	 * if available use a local cache that survives requests, otherwise
+	 * fallback to an in memory implementation of \OCP\ICache
+	 */
+	static private function setupAppInfoCache(){
```
Comment (VicDeo) [inline lib/private/legacy/app.php]: I don't like caches in legacy classes...

File: lib/private/App/AppManager.php [hunk 3]
```diff
@@ -359,7 +360,11 @@ public function getAppsNeedingUpgrade($ocVersion) {
 	 * @internal
 	 */
 	public function getAppInfo($appId) {
-		$appInfo = \OC_App::getAppInfo($appId);
+		try {
+			$appInfo = \OC_App::getAppInfo($appId);
+		} catch (AppNotFoundException $e) {
+			return null;
```
Comment (VicDeo) [inline lib/private/App/AppManager.php]: Invalidates PHPDoc... Bah, it's already outdated.

File: lib/private/App/InfoParser.php [hunk 4]
```diff
@@ -43,12 +47,12 @@ public function parse($file) {
 		libxml_disable_entity_loader($loadEntities);
 		if ($xml === false) {
 			libxml_clear_errors();
-			return null;
+			throw new \InvalidArgumentException('Invalid XML');
```
Comment (VicDeo) [inline lib/private/App/InfoParser.php]: Invalidates PHPDoc - no sense in returning null on error if we are moving to an error control based on exceptions

File: lib/private/App/AppManager.php [hunk 5]
```diff
@@ -519,6 +524,7 @@ public function getAppPath($appId) {
 	 *
 	 * @param string $appId
 	 * @return string|false
+	 * @throws AppNotFoundException
```
Comment (VicDeo) [inline lib/private/App/AppManager.php]: Backward incompatible. Should go into interface as well

File: tests/lib/App/InfoParserTest.php [hunk 6]
```diff
@@ -22,23 +22,17 @@ public function setUp() {
...
-	function providesInfoXml() {
-		return [
-			['expected-info.json', 'valid-info.xml'],
-			[null, 'invalid-info.xml'],
-		];
+	/**
+	 * @expectedException \InvalidArgumentException
+	 */
+	public function testParsingInvalidXml() {
+		$this->parser->parse(OC::$SERVERROOT. "/tests/data/app/valid-info.xml");
```
Comment (VicDeo) [inline tests/lib/App/InfoParserTest.php]: the same filename as in `testParsingValidXml`, is this file able to mutate from valid into invalid state?

**Comment by PVince81** [comment]
is this obsoleted by https://github.com/owncloud/core/pull/30457 ? @VicDeo

**Comment by VicDeo** [comment]
> is this obsoleted by #30457

@PVince81 seems so. At least [this] (https://github.com/owncloud/core/pull/30204/commits/3817c14e234e308b7eda8890d4be375187a649cf) 
and [this](https://github.com/owncloud/core/pull/30204/commits/4bb366ef50d8d3b23d8c35e381bcc1b8506139ff)
commits have the same purpose as https://github.com/owncloud/core/pull/30457

[First one](https://github.com/owncloud/core/pull/30204/commits/a0c381daf6d62be8ec3b7fbdb9ea3450e4dfb087) is similar to https://github.com/owncloud/core/pull/30188

In addition it was created before cs-fixer so it conflicts both with cs-fixer and https://github.com/owncloud/core/pull/30457. I think it's better close this PR and move some parts into  https://github.com/owncloud/core/issues/30190 (if needed).

**Comment by VicDeo** [comment]
https://github.com/owncloud/core/issues/30190 is already fixed in master and has a pending backport.

Closed in favor of https://github.com/owncloud/core/pull/30188

**Comment by butonic** [comment]
[as I said](https://github.com/owncloud/core/pull/30204#issue-164077774), this is based on #30188

I used a separate PR to do things incrementally. I'll go for #30188 now

**Pull request closed without merging**